<a href="https://colab.research.google.com/github/stevecoffee/dasc5301/blob/main/p2/dasc5301_fall2021_p2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DASC5301 Data Science, Fall 2021, Chengkai Li, Unversity of Texas at Arlington
# Programming Assignment 2
# Due: Friday, Octoer 29, 2021, 11:59pm

## **Requirements**

1.   When you work on this assignment, you should make a copy of this notebook in Google Colab. This can be done using the option `File > Save a copy in Drive` in Google Colab. 
2.   You should fill in your answer for each task inside the code block right under the task. 
3.   You should only insert your code into the designated code blocks, as mentioned above. Other than that, you shouldn't change anything else in the notebook.
4.   The correct output for each task is also included right below the corresponding code block. 
5.   Each task can be solved using one line of code. (An exception is Tasks 12, for which it is much easier to use three lines.) Nevertheless, for any task, if you have to use up to three lines, that's fine. But you are not allowed to use more than three lines of code. Note that we have two designated code blocks in both Task 3 and Task 4. 
6.   You may not use any other imports to solve the tasks. In other words, you shouldn't use `import` in any desinated code blocks for the tasks.
7.   You should not use any loops, if statements, or list/dictionary comprehensions. You can solve all the tasks by only using features and functions from pandas. 
8.   Even if you can only partially solve a task, you should include your code in the code block, which allows us to consider partial credit. 
9.   However, your code should not raise errors. Any code raising errors will not get partial credit. 
10.   We tried to minimize the interdependence of the tasks. In most cases, even if you don't solve a task, it won't affect you working on the tasks afterwards, although the output of a task may not be fully correct if you didn't correctly solve the preceding tasks. If you get stuck on a task, you can move on to work on the rest and try to come back to that task later.  
11.   To submit your assignment, download your Colab into a .ipynb file. This can be done using the option `Download > Download .ipynb` in Google Colab.
12.   Submit the downloaded .ipynb file into the Programming Assignment 2 entry in Canvas.

## **Dataset**

In this assignment, we will do data munging and analysis on a dataset about board games. The dataset is from Kaggle, at https://www.kaggle.com/andrewmvd/board-games. We have already downloaded the CSV file and provided the file in the assignment's entry in canvas. You will need to upload the CSV file to your Google Colab working directory. After that, let's load the CSV file into a pandas DataFrame.

In [ ]:
import pandas as pd

games = pd.read_csv('bgg_dataset.csv', delimiter=';', decimal=",")

We set `display.max_rows` to `None` for now so that we can see more values in code output. 

In [ ]:
pd.set_option('display.max_rows', None)

Let's gain some basic understanding of the dataset by using `info()`, `head()`, and `describe()`.

In [ ]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20343 entries, 0 to 20342
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  20327 non-null  float64
 1   Name                20343 non-null  object 
 2   Year Published      20342 non-null  float64
 3   Min Players         20343 non-null  int64  
 4   Max Players         20343 non-null  int64  
 5   Play Time           20343 non-null  int64  
 6   Min Age             20343 non-null  int64  
 7   Users Rated         20343 non-null  int64  
 8   Rating Average      20343 non-null  float64
 9   BGG Rank            20343 non-null  int64  
 10  Complexity Average  20343 non-null  float64
 11  Owned Users         20320 non-null  float64
 12  Mechanics           18745 non-null  object 
 13  Domains             10184 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 2.2+ MB


In [ ]:
games.head()

ID  ...                         Domains
0  174430.0  ...  Strategy Games, Thematic Games
1  161936.0  ...  Strategy Games, Thematic Games
2  224517.0  ...                  Strategy Games
3  167791.0  ...                  Strategy Games
4  233078.0  ...  Strategy Games, Thematic Games

[5 rows x 14 columns]

In [ ]:
games.describe()

ID  Year Published  ...  Complexity Average    Owned Users
count   20327.000000    20342.000000  ...        20343.000000   20320.000000
mean   108216.245142     1984.249877  ...            1.991188    1408.457628
std     98682.097298      214.003181  ...            0.848903    5040.179315
min         1.000000    -3500.000000  ...            0.000000       0.000000
25%     11029.000000     2001.000000  ...            1.330000     146.000000
50%     88931.000000     2011.000000  ...            1.970000     309.000000
75%    192939.500000     2016.000000  ...            2.540000     864.000000
max    331787.000000     2022.000000  ...            5.000000  155312.000000

[8 rows x 11 columns]

## **Need for data cleaning and preprocessing**

The results of these several functions indicate a few needs for cleaning and preprocessing the data:

1) The columns `ID`, `Year Published` and `Owned Users` should be integers, but they are floating point numbers.

2) There are null values in various columns. 

3) It seems there could be wrong values. For instance, the minimal value in `Max Players` is 0. What kind of game is that if it allows at most zero player? 

4) The values in columns `Mechanics` and `Domains` are comma-separated lists. We need to parse these values and get the individual items from the lists.


Let's find out which columns have null values. This could be derived from the `Non-Null Count` in the output of `games.info()`. But there are much simpler ways.

## **Task 1: For each column, find the number of rows with null value in that column. (5 points)** 

If your code is correct, its output should tell you that five columns have null values --- column `ID` has missing value in 16 rows, `Year Published` has 1, `Owned Users` has 23, `Mechanics` has 1598, and `Domains` has 10159! Other columns have no null values. 

In [ ]:
# Code for Task 1


ID                       16
Name                      0
Year Published            1
Min Players               0
Max Players               0
Play Time                 0
Min Age                   0
Users Rated               0
Rating Average            0
BGG Rank                  0
Complexity Average        0
Owned Users              23
Mechanics              1598
Domains               10159
dtype: int64

Since column `ID` has null values, it couldn't be used for uniquely identifying games. Hence, let's take it out. 

## **Task 2: Remove the `ID` column from the `DataFrame` `games`. (5 points)**

In [ ]:
# Code for Task 2


## **Task 3: Replace null values in column `Year Published` by 5000. Replace null values in column `Owned Users` by -1. Note that this task has two desiganated code blocks. (5 points)** 

The column `Year Published` has negative values, as `games.descrie()` shows. It actually has 0 in its values too. Hence, we are using a year in the future (5000) to indicate the dataset doesn't provide the value for a game. The column `Owned Users` has 0s too. We thus use -1 to indicate missing values.

In [ ]:
# Code for Task 3 : code block for replaceing null values in column Year Published by 5000.


In [ ]:
# Code for Task 3: code block for replacing null values in column Owned Users by -1.


## **Task 4: Convert the data type of column ``Year Pubblished`` to integer. Convert the data type of column ``Owned Users`` to integer too.  Note that this task has two desiganated code blocks. (5 points)**

In [ ]:
# Code for Task 4: code block for converting the data type of column ``Year Pubblished`` to integer.


In [ ]:
# Code for Task 4: code block for converting the data type of column ``Owned Users`` to integer. 


After you finish Tasks 1 to 4, run `games.info()`, `games.head()`, and `games.describe()` again, to verify you have achieved the goals. In fact, you could do this from time to time, in various places, to make sure you haven't messed up the data.

In [ ]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20343 entries, 0 to 20342
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Name                20343 non-null  object 
 1   Year Published      20343 non-null  int16  
 2   Min Players         20343 non-null  int64  
 3   Max Players         20343 non-null  int64  
 4   Play Time           20343 non-null  int64  
 5   Min Age             20343 non-null  int64  
 6   Users Rated         20343 non-null  int64  
 7   Rating Average      20343 non-null  float64
 8   BGG Rank            20343 non-null  int64  
 9   Complexity Average  20343 non-null  float64
 10  Owned Users         20343 non-null  int32  
 11  Mechanics           18745 non-null  object 
 12  Domains             10184 non-null  object 
dtypes: float64(2), int16(1), int32(1), int64(6), object(3)
memory usage: 1.8+ MB


In [ ]:
games.head()

Name  ...                         Domains
0                         Gloomhaven  ...  Strategy Games, Thematic Games
1          Pandemic Legacy: Season 1  ...  Strategy Games, Thematic Games
2                  Brass: Birmingham  ...                  Strategy Games
3                  Terraforming Mars  ...                  Strategy Games
4  Twilight Imperium: Fourth Edition  ...  Strategy Games, Thematic Games

[5 rows x 13 columns]

In [ ]:
games.describe()

Year Published   Min Players  ...  Complexity Average    Owned Users
count    20343.000000  20343.000000  ...        20343.000000   20343.000000
mean      1984.398122      2.019712  ...            1.991188    1406.864081
std        215.039951      0.690366  ...            0.848903    5037.551824
min      -3500.000000      0.000000  ...            0.000000      -1.000000
25%       2001.000000      2.000000  ...            1.330000     146.000000
50%       2011.000000      2.000000  ...            1.970000     308.000000
75%       2016.000000      2.000000  ...            2.540000     864.000000
max       5000.000000     10.000000  ...            5.000000  155312.000000

[8 rows x 10 columns]

Earlier we noticed the existence of value 0 in certain columns which shouldn't have such values. Let's find out how prevalent the prolem is. 

## **Task 5: For each column, show how many rows have 0 as the value. (5 points)**



In [ ]:
# Code for Task 5


Name                     0
Year Published         185
Min Players             46
Max Players            161
Play Time              556
Min Age               1251
Users Rated              0
Rating Average           0
BGG Rank                 0
Complexity Average     426
Owned Users              1
Mechanics                0
Domains                  0
dtype: int64

If you get the correct code, you will see from the output that 46 rows have value 0 in column `Min Players`, 161 in column `Max Players`, 185 in `Year Published`, and so on. We need to keep this mind when we analyze the data so that we don't draw inaccurate conclusions. 

Particularly, let's examine `Year Published`. We discovered earlier that it also has negative values. We can take a closer look now.




## **Task 6: Get the number of games published in each year. Sort the years by frequency, in descending order. (10 points)**

If you get the code correct, you will find that Year 2017 has 1274 games published, which is the most among all years. 

In [ ]:
# Code for Task 6


Year Published
 2017    1274
 2016    1257
 2018    1254
 2019    1134
 2015    1131
 2014     987
 2013     850
 2012     815
 2011     735
 2010     692
 2020     684
 2009     631
 2008     580
 2005     544
 2007     522
 2006     516
 2004     490
 2003     408
 2002     335
 2001     298
 2000     295
 1999     268
 1998     246
 1997     221
 1996     216
 1995     212
 1992     201
 0        185
 1993     183
 1994     180
 1991     175
 1981     149
 1986     149
 1990     146
 2021     144
 1987     136
 1989     132
 1983     130
 1985     129
 1988     128
 1979     126
 1980     125
 1982     124
 1975     113
 1977     102
 1978      97
 1984      96
 1973      73
 1974      69
 1972      66
 1976      64
 1971      36
 1970      33
 1965      26
 1969      25
 1968      24
 1967      20
 1960      19
 1964      19
 1962      18
 1963      17
 1961      11
 1930      11
 1959      11
 1966      10
 1956       9
 1958       9
 1954       7
 1935       7
 1940       6
 1938

The oldest game was from 3500 BC. Let's find out which game it is. 

## **Task 7: Find the name of the oldest game, based on column `Year Published`. (5 points)**

If your code is correct, it will return Senet as the name of the oldest game. Some Googling wnd Wikipediaing will verify this game is indeed ancient, although we couldn't verify the accuracy of the exact value -3500. That's fine.

In [ ]:
# Code for Task 7


'Senet'

In the output of Task 6, you see that in general the number of games published in a year has been steadily increasing. However, Year 0 appears to be an outlier, as it has 185 games according to the dataset. This doesn't seem right. To further verify, let's do the following Task 8.

## **Task 8: Find how many games in total have been publisehd before 1900. Do not include Year 0 in the count. (5 points)**

If your code is correct, it shall return 111. We can thus conclude the statistics for Year 0 cannot be correct. Most likely this is because 0 was used to indicate unknown/missing publishing year when the dataset was created. How confusing that is. This reminds us it is important to make good choices in dealing with missing values. 

In [ ]:
# Code for Task 8


111

For the same reason, we believe value 0 in all other columns are not reliable either. Our next task will replace 5000 in `Year Published` and -1 in `Owned Users` by 0. Remember they were actually null values (Task 3). Later we will ignore them together with all 0 values in our analysis.

## **Task 9: Replace 5000 in `Year Published` and -1 in `Owned Users` by 0. (5 points)**

In [ ]:
# Code for Task 9


Also, let's take a look at the frequency of each value in column `Min Players`. If you write a piece of code to find out, the code will be similiar to the one in Task 6. Hence, we don't provide it here. Instead, we directly provide the values, as follows. 

In [ ]:
pd.Series({0: 46, 1: 3270, 2: 14076, 3: 2365, 4: 474, 5: 57, 6: 21, 7: 14, 8: 17, 9: 1, 10: 2}, name = 'Min Players')

0        46
1      3270
2     14076
3      2365
4       474
5        57
6        21
7        14
8        17
9         1
10        2
Name: Min Players, dtype: int64

Our next task attempts to examine the relationship between `Min Players` and popularity of games measured by ownership. 

## **Task 10: For each value of `Min Players`, find the average `Owned Users` for games with the corresponding `Min Players` value. Exclude the games with values 0, 9, 10 in `Min Players` and value 0 in `Owned Users`. (10 points)**

As we discussed in Task 8, we don't trust the value 0 in any of the columns. We can thus ignore the games with 0 in `Min Players`. Furthermore, there are only 1 and 2 games for `Min Players` 9 and 10, respectively. The statistics of these games won't be meaningful. When we focus on the rest of the games in the output of Task 10, we will observe a general pattern of decresing ownership by minimum required players. That is probably not surprising, since it is easier to find people to play games with less required players.  

In [ ]:
# Code for Task 10


Min Players
1    1762.693909
2    1348.215520
3    1309.574756
4    1218.301688
5    2827.526316
6    1154.761905
7     332.571429
8    1661.705882
Name: Owned Users, dtype: float64

The pattern from Task 10 appears to have some exceptions, in games with `Min Players` being 5 and 8. Let's take a further look. Before we continue, let's change `display.max_rows` to 50.

In [ ]:
pd.set_option('display.max_rows', 50)

## **Task 11: Find the top-5 owned games for each group of games based on `Min Players`. Ideally, you should also exclude the games with values 0, 9, 10 in `Min Players`, like what we did in Task 10. It is fine if you don't do it here. (10 points)**

From the output, you see that there are some quite popular games with 5 and 8 `Min Players`.

In [ ]:
# Code for Task 11 


Name  ...                         Domains
98                                           Pandemic  ...    Family Games, Strategy Games
394                                             Catan  ...    Family Games, Strategy Games
177                                       Carcassonne  ...                    Family Games
60                                          7 Wonders  ...    Family Games, Strategy Games
92                                          Codenames  ...                     Party Games
97                                           Dominion  ...                  Strategy Games
3                                   Terraforming Mars  ...                  Strategy Games
30                                           Agricola  ...                  Strategy Games
13                                             Scythe  ...                  Strategy Games
4401                                         Munchkin  ...                  Thematic Games
27                                        Puerto Rico  ...                  Strategy Games
277                                             Dixit  ...                     Party Games
0                                          Gloomhaven  ...  Strategy Games, Thematic Games
19                                           Wingspan  ...    Family Games, Strategy Games
518                     Betrayal at House on the Hill  ...                  Thematic Games
271                                    The Resistance  ...                     Party Games
4300                           Cards Against Humanity  ...                     Party Games
130                            The Resistance: Avalon  ...                     Party Games
1263                                            BANG!  ...                     Party Games
4786                                 Apples to Apples  ...                     Party Games
185                                     Secret Hitler  ...                     Party Games
759                                           Concept  ...                     Party Games
192                    Deception: Murder in Hong Kong  ...                     Party Games
1012                The Werewolves of Miller's Hollow  ...                     Party Games
655                    A Fake Artist Goes to New York  ...                     Party Games
903                              Two Rooms and a Boom  ...                     Party Games
1602                                    Mafia de Cuba  ...                     Party Games
973               Ultimate Werewolf: Ultimate Edition  ...                     Party Games
1663                                         Werewolf  ...                     Party Games
2759                                        Dead Last  ...                     Party Games
2358                                      Blood Bound  ...                     Party Games
3540                                  Lupus in Tabula  ...                     Party Games
2650   The Werewolves of Miller's Hollow: The Village  ...                     Party Games
19730                                Templar Intrigue  ...                             NaN
18574                                The Last Banquet  ...                     Party Games
5507                               Ca$h 'n Gun$: Live  ...                     Party Games
12008                                 Mafia: Vendetta  ...                             NaN
14924                                           Mafia  ...                     Party Games
7682                                       Witch Hunt  ...                     Party Games
9913                                    Fox & Chicken  ...                             NaN

[40 rows x 13 columns]

## **Task 12: Produce a pivot table using `Rating Average` as the rows (i.e., index) and `Complexity Average` as the columns. The cells of the pivot table should show the number of games having the corresponding rating average and complexity average. Since both `Rating Average` and `Complexity Average` are floating point numbers, we should use bins on both. Let's create ten bins [1,2], (2,3], (3,4], (4,5], (5,6], (6,7], (7,8], (8,9], (9,10] on `Rating Average` and four bins [1,2], (2,3], (3,4], (4,5] on `Complexity Average`. (10 points)**

The output of `games.describe()` shows that column `Complexity Average` has value 0. Based on our earlier analysis, we shouldn't put much faith in this value. Besides 0, the smallest value in that column is 1. Therefore we decide to have four bins for `Complexity Average`. 

In [ ]:
# Code for Task 12


Complexity Average  (0.999, 2.0]  (2.0, 3.0]  (3.0, 4.0]  (4.0, 5.0]
Rating Average                                                      
(0.999, 2.0]                 6.0         2.0         NaN         NaN
(2.0, 3.0]                  19.0         4.0         1.0         NaN
(3.0, 4.0]                 154.0        16.0         3.0         1.0
(4.0, 5.0]                1023.0       101.0        18.0         2.0
(5.0, 6.0]                3972.0       860.0       127.0         8.0
(6.0, 7.0]                4856.0      2906.0       650.0        57.0
(7.0, 8.0]                1209.0      1978.0      1029.0       163.0
(8.0, 9.0]                 102.0       276.0       256.0        88.0
(9.0, 10.0]                  9.0         9.0         9.0         3.0

The pivot table suggests a positive correlation between these two columns. As the complexity of games increases, the rating also tends to increase. Perhaps this is intuitive. For complex games to have a market, it needs to be of higher quality. 

In fact, we can directly calculate the correlation using `corr`, as follows. The value of 0.5 in Pearson correlation coefficient suggests a fairly large positive correlation.

In [ ]:
g = games[games['Complexity Average']>0]

g['Rating Average'].corr(g['Complexity Average'], method='pearson')

0.510870853513205

In the next task, we are going to produce a similar pivot table, focusing on `Owned Users`. Ideally we want to exclude the games with value 0 on this column. To simplify things, we are not requiring you to do it. The pattern we will be seeing is not changed, since only 23 games would have been removed. 

## **Task 13: Produce another pivot table using `Rating Average` and `Complexity Average`, with the same binning. However, this time the cells of the pivot table should show average `Owned Users` of games matching the corresponding rating average and complexity average. (5 points)**

In [ ]:
# Code for Task 13


Complexity Average  (0.999, 2.0]   (2.0, 3.0]   (3.0, 4.0]   (4.0, 5.0]
Rating Average                                                         
(0.999, 2.0]           21.000000    90.000000          NaN          NaN
(2.0, 3.0]            526.736842   119.750000   133.000000          NaN
(3.0, 4.0]            364.188312   234.125000   207.666667   176.000000
(4.0, 5.0]            436.129032   232.287129   240.444444    71.000000
(5.0, 6.0]            539.152568   485.956977   388.976378   321.000000
(6.0, 7.0]           1150.022446  1091.727116   885.712308   811.631579
(7.0, 8.0]           3296.243176  3273.025784  2928.022352  1832.558282
(8.0, 9.0]            734.127451  2583.463768  4430.777344  3669.193182
(9.0, 10.0]            62.222222    66.555556    75.666667   196.000000

If your code is correct, this pivot table also shows some interesting patterns. At every rating tier till (7, 8], simpler games enjoy larger ownerships. However, for the really good games with ratings greater than 8, players are not afraid of their complexity. In fact, the more complex games in this rating tier get owned by more players.

Now we will process the `Mechanics` and `Domains` columns. They store values as strings. Each string is a comma-separated list of items. The following code will turn `Mechanics` into a DataFrame itself, with each column corresponding to a unique item from the comma-separated lists. Similarly, we are creating a new DataFrame for the `Domains` column.

In [ ]:
mechanics = games['Mechanics'].str.get_dummies(sep=", ")
domains = games['Domains'].str.get_dummies(sep=", ")

mechanics = pd.concat([games['Name'], mechanics], 1)
domains = pd.concat([games['Name'], domains], 1)

Let's take a look at the columns in the new DataFrame `mechanics`. The values are 1 and 0, i.e., essentially Boolean, indicating whether a game uses the corresponding mechanics or not. This is also called *one-hot encoding*. 


In [ ]:
mechanics.head()

Name  ...  Zone of Control
0                         Gloomhaven  ...                0
1          Pandemic Legacy: Season 1  ...                0
2                  Brass: Birmingham  ...                0
3                  Terraforming Mars  ...                0
4  Twilight Imperium: Fourth Edition  ...                0

[5 rows x 183 columns]

Similarly the new DataFrame `domains` uses one-hot encoding to record the games' domain types. 

In [ ]:
domains.head()

Name  Abstract Games  ...  Thematic Games  Wargames
0                         Gloomhaven               0  ...               1         0
1          Pandemic Legacy: Season 1               0  ...               1         0
2                  Brass: Birmingham               0  ...               0         0
3                  Terraforming Mars               0  ...               0         0
4  Twilight Imperium: Fourth Edition               0  ...               1         0

[5 rows x 9 columns]

Let's find out which are the most common game domains and which are the least common ones. If your code for the following task is correct, the output should show there are 3316 `Wargames`, the most common type. The least common type is `Customizable Games`, with 297 games. 

## **Task 14: For each domain, list its frequency, i.e., the number of games belonging to that domain. (5 points)**

In [ ]:
# Code for Task 14


Abstract Games        1070
Children's Games       849
Customizable Games     297
Family Games          2173
Party Games            605
Strategy Games        2205
Thematic Games        1174
Wargames              3316
dtype: int64

## **Task 15: Find out the average `Complexity Average` of games that belong to `Wargames`. You should exclude games with 0 on `Complexity Average`. (10 points)**

In [ ]:
# Code for Task 15 


2.8743964996982436

If you code is correct, 2.874 is the value. If you perform the same task on `Children's Games`, you will get 1.175. These two are the two types of games with the largest and smallest average `Complexity Average`. If you want to calculate this for every type of games, the following code does it.

In [ ]:
results = \
    (games[games['Complexity Average']>0].set_index(games.columns.drop('Domains',1).tolist())
    .Domains.str.split(', ', expand=True)
    .stack()
    .reset_index()
    .rename(columns={0:'domain'})
    .loc[:,['domain','Owned Users', 'Rating Average', 'Complexity Average']]
    .groupby('domain').agg({'Owned Users':['mean'], 'Rating Average':['mean'], 'Complexity Average':['mean']})
    )
results 

Owned Users Rating Average Complexity Average
                           mean           mean               mean
domain                                                           
Abstract Games      1362.797170       6.228330           1.968028
Children's Games     733.255675       5.516237           1.175042
Customizable Games  2247.594595       6.365473           2.408581
Family Games        4071.729867       6.405568           1.658739
Party Games         4089.066116       6.285752           1.346545
Strategy Games      4438.449433       6.957982           2.710018
Thematic Games      4045.533220       6.789233           2.441235
Wargames             705.910984       6.847182           2.874396